packages/dds/tree/src/simple-tree/tree.ts

```diff
@@ -140,18 +161,117 @@ export class TreeViewConfiguration<TSchema extends ImplicitFieldSchema = Implici
 		this.schema = config.schema;
 		this.enableSchemaValidation = config.enableSchemaValidation;
 
-		// Ensure all reachable schema are marked as most derived.
-		// This ensures if multiple schema extending the same schema factory generated class are present (or have been constructed, or get constructed in the future),
-		// an error is reported.
-		walkFieldSchema(config.schema, markSchemaMostDerived);
+		// Ambiguity errors are lower priority to report than invalid schema errors, so collect these in an array and report them all at once.
+		const ambiguityErrors: string[] = [];
+
+		walkFieldSchema(config.schema, {
+			// Ensure all reachable schema are marked as most derived.
+			// This ensures if multiple schema extending the same schema factory generated class are present (or have been constructed, or get constructed in the future),
+			// an error is reported.
+
+			node: markSchemaMostDerived,
+			allowedTypes(types): void {
+				if (config.ensureUnambiguous) {
+					checkUnion(types, ambiguityErrors);
+				}
+			},
+		});
+
+		if (ambiguityErrors.length !== 0) {
+			throw new UsageError(`Ambagious Schema Found:\n${ambiguityErrors.join("\n")}`);
+		}
+
 		// Eagerly perform this conversion to surface errors sooner.
 		toFlexSchema(config.schema);
 	}
 }
 
+function checkUnion(union: Iterable<TreeNodeSchema>, errors: string[]): void {
+	const checked: Set<TreeNodeSchema> = new Set();
+	const maps: TreeNodeSchema[] = [];
+	const arrays: TreeNodeSchema[] = [];
+
+	const objects: ObjectNodeSchema[] = [];
+	// Map from key to schema using that key
+	const allObjectKeys: Map<string, ObjectNodeSchema[]> = new Map();
+
+	for (const schema of union) {
+		if (checked.has(schema)) {
+			throw new UsageError(`Duplicate schema in allowed types: ${schema.identifier}`);
+		}
+		checked.add(schema);
+
+		if (union instanceof LeafNodeSchema) {
```

Comment:
This should be `schema`, rather than `union`, right?

Reply:
yes.
```suggestion
		if (schema instanceof LeafNodeSchema) {
```